File: AVLTree/src/AVLTree.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

// AVL平衡二叉树，防止之前的二分搜索树退化为链表，所有操作为O(logn)级别的
public class AVLTree<K extends Comparable<K>, V>{

    private class Node{
        public K key;
        public V value;
        public Node left, right;
        public int height;   // 记录每个节点的高度

        public Node(K key, V value){
            this.key = key;
            this.value = value;
            left = null;
            right = null;
            height = 1;
        }
    }

    private Node root;
    private int size;

    public AVLTree(){
        root = null;
        size = 0;
    }

    
    public int getSize() {
        return size;
    }
    
    public boolean isEmpty() {
        return size == 0;
    }

    // 判断该二叉树是否是一颗二分搜索树
    public boolean isBST(){

        ArrayList<K> keys = new ArrayList<>();
        inOrder(root, keys);

        // 进行中序遍历之后检查存入到keys数组中的元素是不是升序
        for (int i = 1; i < keys.size(); i ++){
            if (keys.get(i - 1).compareTo(keys.get(i)) > 0 )
                return false;
        }

        return true;
    }

    // 二分搜索树的中序遍历,将中序遍历的结果存入keys数组中
    private void inOrder(Node node, ArrayList<K> keys){

        if (node == null)
            return;

        inOrder(node.left, keys);
        keys.add(node.key);
        inOrder(node.right, keys);
    }

    // 二分搜索树的中序遍历,将中序遍历的结果输出
    public void inOrderR(){
        inOrderR(root);
    }

    private void inOrderR(Node node){

        if (node == null)
            return;

        inOrderR(node.left);
        System.out.print(node.value + " ");
        inOrderR(node.right);
    }

    // 二分搜索树的层序遍历
    public void levelOrder(){

        Queue<AVLTree.Node> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            AVLTree.Node cur = q.remove();
            System.out.print(cur.key + " ");

            if(cur.left != null)
                q.add(cur.left);
            if(cur.right != null)
                q.add(cur.right);
        }
    }

    // 判断该二叉树是否是一颗平衡二叉树
    public boolean isBalanced(){
        return isBalanced(root);
    }

    // 判断以Node为根的二叉树是否是一颗平衡二叉树，递归算法
    private boolean isBalanced(Node node){

        if (node == null)
            return true;

        int balanceFactor = getBalanceFactor(node);
        if (Math.abs(balanceFactor) > 1)
            return false;

        return isBalanced(node.left) && isBalanced(node.right);
    }

    // 获得节点node的高度
    private int getHeight(Node node){
        if (node == null)
            return 0;

        return node.height;
    }

    // 获得节点node的平衡因子
    // 左子树高度 - 右子树高度
    private int getBalanceFactor(Node node){
        if (node == null)
            return 0;
        return getHeight(node.left) - getHeight(node.right);
    }

    /*
    对y节点进行向右旋转的操作，返回旋转后新的根节点
                 y                              x
                / \                          /     \
               x    T4   向右旋转（y）        z       y
              / \       ---------->       /  \      /  \
             z   T3                      T1  T2    T3  T4
            / \
           T1  T2
     在y这个节点不满足平衡性
     */
    private Node rightRotate(Node y){

        Node x = y.left;
        Node T3 = x.right;

        // 向右旋转的过程
        x.right = y;
        y.left = T3;

        // 更新height
        //只有x和y节点的高度发生变化
        y.height = Math.max(getHeight(y.left), getHeight(y.right)) + 1;
        x.height = Math.max(getHeight(x.left), getHeight(x.right)) + 1;

        return x;
    }

    /*
    对y节点有进行向左旋转，返回旋转后新的根节点x
                 y                                    x
                / \                                  / \
               T4   x         向左旋转（y）           y    z
                   / \       ---------->          / \   / \
                  T3  z                          T4 T3 T1 T2
                     / \
                   T1  T2


     */

    private Node leftRotate(Node y){

        Node x = y.right;
        Node T2 = x.left;

        // 向左旋转过程
        x.left = y;
        y.right = T2;

        // 更新Height值
        y.height = Math.max(getHeight(y.left), getHeight(y.right)) + 1;
        x.height = Math.max(getHeight(x.left), getHeight(x.right)) + 1;

        return x;

    }

    // 向AVL树中添加新的元素（key， value）
    public void add(K key, V value) {
        root = add(root, key, value);
    }

    // 向以node为根的二分搜索树中插入元素（key， value），递归算法
    // 返回插入新节点后二分搜索树的根
    private Node add(Node node, K key, V value){

        if(node == null){
            size ++;
            return new Node(key, value);
        }

        // 递归过程， 典型二分搜索树添加过程
        if(key.compareTo(node.key) < 0)
            node.left = add(node.left, key, value);
        else if(key.compareTo(node.key) > 0)
            node.right = add(node.right, key, value);
        else    // key.compareTo(node.key) == 0
            node.value = value;

        // 更新height
        node.height = 1 + Math.max(getHeight(node.left), getHeight(node.right));
        // 计算平衡因子
        int balanceFactor = getBalanceFactor(node);

//        System.out.println("rotate before unbalanced : " + balanceFactor);
//        if (Math.abs(balanceFactor) > 1){
//            System.out.println("unbalanced : " + balanceFactor);
//        }


        // 平衡维护
        // 左节点的高度大于右节点高度，且这个不平衡性是由于该节点左侧的左侧多添加了
        // 一个节点
        // LL
        if (balanceFactor > 1 && getBalanceFactor(node.left) >= 0){
            // 使用右旋转，使这棵树进行平衡
            // 返回新的根节点
            return rightRotate(node);
        }

        // 平衡维护
        // 左节点的高度大于右节点高度，且这个不平衡性是由于该节点右侧的右侧多
        // 添加了一个节点
        // RR
        if (balanceFactor < -1 && getBalanceFactor(node.right) <= 0){
            // 使用左旋转，使这棵树进行平衡
            // 返回新的根节点
            return leftRotate(node);
        }

        // LR
        if (balanceFactor > 1 && getBalanceFactor(node.left) < 0){
            node.left = leftRotate(node.left);
            return rightRotate(node);
        }

        //RL
        if (balanceFactor < -1 && getBalanceFactor(node.right) > 0){
            node.right = rightRotate(node.right);
            return leftRotate(node);
        }

//        System.out.println('(' + "node.key:" + node.key + ',' +
//                        "node.value:" +  node.value + ')');
//        System.out.println("rotate after unbalanced : " + balanceFactor);
//        System.out.println('\n');
        return node;
    }

    // 返回以node为根节点的二分搜索树中，key所在的节点
    private Node getNode(Node node, K key){

        if (node == null)
            return null;

        if (key.compareTo(node.key) == 0)
            return node;
        else if (key.compareTo(node.key) < 0)
            return getNode(node.left, key);
        else    // if (key.compareTo(node.key)) > 0
            return getNode(node.right, key);
    }
    
    public boolean contains(K key) {
        return getNode(root, key) != null;
    }

    
    public V get(K key) {
        Node node = getNode(root, key);

        return node == null ? null : node.value;
    }

    // 更新操作
    public void set(K key, V newValue) {
        Node node = getNode(root, key);

        if (node == null)
            throw new IllegalArgumentException(key  + " doesn`t exists");

        node.value = newValue;
    }

    // 返回以node为根的二分搜索树的最小值所在的节点
    private Node minimum(Node node){
        if(node.left == null)
            return node;
        return minimum(node.left);
    }

//    // 删除掉以node为根的二分搜索树中的最小节点
//    // 返回删除节点后新的二分搜索树的根
//    private Node removeMin(Node node){
//
//        if(node.left == null){
//            Node rightNode = node.right;
//            node.right = null;
//            size --;
//            return rightNode;
//        }
//
//        node.left = removeMin(node.left);
//        return node;
//    }


    // 从二分搜索树中删除键为key的节点
    public V remove(K key) {

        Node node = getNode(root, key);
        if (node != null){
            root = remove(root, key);
            return node.value;
        }

        return null;
    }


    // 删除以node为根的二分搜索树中键为key的节点，递归算法
    // 返回删除节点后新的二分搜索树的根
    private Node remove(Node node, K key){

        if (node == null){
            return null;
        }

        Node retNode;   // 暂存返回的新子树根节点

        if (key.compareTo(node.key) < 0){
            node.left = remove(node.left, key);
            retNode = node;
        }
        else if (key.compareTo(node.key) > 0){
            node.right = remove(node.right, key);
            retNode = node;
        }
        else{   // key.compareTo(node.key) == 0

            if (node.left == null) {
                // 待删除节点左子树为空的情况
                // 此时的node为待删除节点
                Node rightNode = node.right;
                node.right = null;
                size--;
                retNode = rightNode;   // 返回右子树根节点
            }
            else if (node.right == null){
                // 待删除节点右子树为空的情况
                Node leftNode = node.left;
                node.left = null;
                size --;
                retNode = leftNode;
            }
            else{

                // 待删除节点左右子树都不为空的情况
                // 找到比待删除节点大的节点, 即待删除节点右子树的最小节点
                // 或比待删除结点小的节点，即左子树最大节点
                // 用这个节点顶替待删除节点的位置
                Node successor = minimum(node.right);
                successor.right = remove(node.right, successor.key);
                successor.left = node.left;

                node.left = node.right = null;

                retNode = successor;
            }
        }

        if (retNode == null)
            return null;

        /*
        自平衡处理
         */

        // 更新height
        retNode.height = 1 + Math.max(getHeight(retNode.left), getHeight(retNode.right));
        // 计算平衡因子
        int balanceFactor = getBalanceFactor(retNode);


        // 平衡维护
        // 左节点的高度大于右节点高度，且这个不平衡性是由于该节点左侧的左侧多添加了
        // 一个节点
        // LL
        if (balanceFactor > 1 && getBalanceFactor(retNode.left) >= 0){
            // 使用右旋转，使这棵树进行平衡
            // 返回新的根节点
            return rightRotate(retNode);
        }

        // 平衡维护
        // 左节点的高度大于右节点高度，且这个不平衡性是由于该节点右侧的右侧多
        // 添加了一个节点
        // RR
        if (balanceFactor < -1 && getBalanceFactor(retNode.right) <= 0){
            // 使用左旋转，使这棵树进行平衡
            // 返回新的根节点
            return leftRotate(retNode);
        }

        // LR
        if (balanceFactor > 1 && getBalanceFactor(retNode.left) < 0){
            retNode.left = leftRotate(retNode.left);
            return rightRotate(retNode);
        }

        //RL
        if (balanceFactor < -1 && getBalanceFactor(retNode.right) > 0){
            retNode.right = rightRotate(retNode.right);
            return leftRotate(retNode);
        }

        return retNode;
    }

    public static void main(String[] args) {

        AVLTree<Integer, Integer> avl = new AVLTree<>();
        for (int i = 30; i > 10; i --){
            avl.add(i, i);
        }
        avl.inOrderR();
        System.out.println(" ");
        System.out.println("is BST: " + avl.isBST());
        System.out.println("is Balanced: " + avl.isBalanced());
        avl.levelOrder();

        // write your code here

//        System.out.println("PAP.txt");
//
//        ArrayList<String> words = new ArrayList<>();
//
//        if (FileOperation.readFile("E:/java/AVLTree/PAP.txt", words)){
//
//            System.out.println("Total words: " + words.size());
//
//            AVLTree<String, Integer> map = new AVLTree<>();
//
//            for (String word: words){
//                if (map.contains(word))
//                    map.set(word, map.get(word) + 1);
//                else
//                    map.add(word, 1);
//            }
//            System.out.println("Total different words: " + map.getSize());
//            System.out.println("Frequency of F : " + map.get("r") );
//
//            System.out.println("is BST: " + map.isBST());
//            System.out.println("is balanced: " + map.isBalanced());
//
//            for (String word : words){
//                map.remove(word);
//                System.out.println(word);
//                if (!map.isBST() || !map.isBalanced())
//                    throw new RuntimeException("error");
//            }
//        }
    }

}

File: AVLTree/AVLTree_py/AVLTree.py
import random


class Node:
    def __init__(self, data = None):
        if data == None:
            self.data = 0
        else:
            self.data = data
        self.height = 0

        self.left = None
        self.right = None


class AVLTree:
    def __init__(self):
        self.root = None
        self.size = 0


    def __getNode(self, data = None):
        if data is None:
            return Node()
        else:
            return Node(data)


    def inOrder(self, Node, keys):
        """
        中序遍历
        :param Node:
        :return:
        """
        if Node is None:
            return
        self.inOrder(Node.left, keys)
        keys.append(Node.data)
        print(Node.data)
        self.inOrder(Node.right, keys)

    def isBST(self):
        """
        判断是不是二分搜索树
        :return:
        """
        keys = []
        self.inOrder(self.root, keys)

        for i in range(1, len(keys)):
            if (keys[i] < keys[i - 1]):
                return False

        # print(keys)
        return True

    def isBalanced(self):
        return self.__isBalanced(self.root)

    def __isBalanced(self, node):
        if node == None:
            return True

        balanceFactor = self.getBalanceFactor(node)
        if balanceFactor > 1:
            return False

        return self.__isBalanced(node.left) and self.__isBalanced(node.right)


    def getHeigt(self, Node):
        """
        获得节点的高度
        :param Node: 节点
        :return: 该节点的高度
        """
        if Node is None:
            return 0

        return Node.height

    def getBalanceFactor(self, node):
        """
        计算这个节点的平衡因子
        :param Node:
        :return: 平衡因子
        """
        if node is None:
            return 0

        return self.getHeigt(node.left) - self.getHeigt(node.right)

    def getNode(self, node, data):

        if node == None:
            return None

        if data == node.data:
            return node
        elif data < node.data:
            return self.getNode(node.left, data)
        else:
            return self.getNode(node.right, data)


    def rightRotate(self, node):
        """
        右旋转
        :param Node:
        :return: 此时的根节点
        """
        x = node.left
        node.left = x.right
        x.right = node

        # 更新height
        node.height = max(self.getHeigt(node.right), self.getHeigt(node.left)) + 1
        x.height = max(self.getHeigt(x.right), self.getHeigt(x.left)) + 1

        return x

    def leftRotate(self, node):
        """
        左旋转
        :param node:
        :return:
        """
        x = node.right
        node.left = x.left
        x.left = node


        # 更新height
        node.height = max(self.getHeigt(node.right), self.getHeigt(node.left)) + 1
        x.height = max(self.getHeigt(x.right), self.getHeigt(x.left)) + 1

    def add(self, data):
        """
        向AVL中添加数据
        :param data:
        :return:
        """
        self.root = self.__add(self.root, data)
        print('------------------------------------------')
        print("data = ", data)
        print('self.root.data = :', self.root.data)
        print('------------------------------------------')

    def __add(self, node, data):

        if node == None:
            self.size += 1
            node = self.__getNode(data)
            return node

        # print(node.left is None)

        if data < node.data:
            print("L <---")
            node.left = self.__add(node.left, data)
        elif data > node.data:
            print("---> R")
            node.right = self.__add(node.right, data)
        else:
            node.data = data

        # 更新height
        node.height = 1 + max(self.getHeigt(node.left), self.getHeigt(node.right))
        # 计算平衡因子
        balanceFactor = self.getBalanceFactor(node)

        # 维护平衡
        # 不平衡节点的左孩子的左孩子处添加了节点
        # LL型
        #         y
        #        /
        #       x
        #      /
        #     z
        if balanceFactor > 1 and self.getBalanceFactor(node.left) >= 0:
            print("LL")
            return self.rightRotate(node)

        # RR型
        # y
        #   \
        #     x
        #       \
        #         z
        if balanceFactor < -1 and self.getBalanceFactor(node.right) <= 0:
            print("RR")
            return self.leftRotate(node)

        # LR 型
        #       y
        #      /
        #     x
        #       \
        #         z
        if balanceFactor > 1 and self.getBalanceFactor(node.left) < 0:
            print("LR")
            node.left = self.leftRotate(node.left)
            return self.rightRotate(node)

        # RL型
        # y
        #   \
        #     x
        #    /
        #   z
        if balanceFactor < -1 and self.getBalanceFactor(node.right) > 0:
            print("RL")
            node.right = self.rightRotate(node.right)
            return self.leftRotate(node)

        return node

    def minium(self, node):
        """
        返回以node为根的AVL的最小值所在的节点
        :param node:
        :return:
        """
        if node.left == None:
            return node

        return self.minium(node.left)

    def maxmium(self, node):
        """
        返回以node为根节点的最大值所在的节点
        :param node:
        :return:
        """
        if node.right == None:
            return node

        return self.maxmium(node.right)


    def remove(self, data):
        node = self.getNode(self.root, data)

        if node != None:
            self.root = self.__remove(self.root, data)
            return node.data

        return None

    def __remove(self, node, data):
        """
         删除以node为根的二分搜索树中键为key的节点，递归算法
         返回删除节点后新的二分搜索树的根
        :param node:
        :param data:
        :return:
        """
        if node == None:
            return None

        retNode = self.__getNode()
        if data < node.data:
            node.left = self.__remove(node.left, data)
            retNode = node
        elif data > node.data:
            node.right = self.__remove(node.right, data)
            retNode = node
        else:
            if node.left == None:
                rightNode = node.right
                node.right = None
                self.size -= 1
                retNode = rightNode

            elif node.right == None:
                leftNode = node.left
                node.left = None
                self.size -= 1
                retNode = leftNode

            else:
                successor = self.minium(node.right)
                successor.right = self.__remove(node.right, successor.data)
                successor.left = node.left


        if retNode == None:
            return None

        retNode.height = 1 + max(self.getHeigt(retNode.left), self.getHeigt(retNode.right))

        balanceFactor = self.getBalanceFactor(retNode)

        if balanceFactor > 1 and self.getBalanceFactor(node.left) >= 0:
            print("LL")
            return self.rightRotate(retNode)

        if balanceFactor < -1 and self.getBalanceFactor(node.right) <= 0:
            print("RR")
            return self.leftRotate(retNode)

        if balanceFactor > 1 and self.getBalanceFactor(node.left) < 0:
            print("LR")
            retNode.left = self.leftRotate(retNode.left)
            return self.rightRotate(retNode)

        if balanceFactor < -1 and self.getBalanceFactor(retNode.right) > 0:
            print("RL")
            retNode.right = self.rightRotate(retNode.right)
            return self.leftRotate(retNode)

        return retNode




if __name__ == "__main__":
    avlTree = AVLTree()
    avlTree.add(1)
    avlTree.add(10)
    avlTree.add(4)
    avlTree.add(5)
    avlTree.add(11)

    avlTree.remove(10)

    print(f"avlTree.isBST = {avlTree.isBST()}")
    print(f"avlTree.isBalanced = {avlTree.isBalanced()}")
